### PR TITLE
python312Packages.mdformat-gfm-alerts: init at 1.0.1

### LINKED ISSUE #328644: Package request: mdformat-gfm-alerts (python package)
**Project description**

An [mdformat](https://github.com/executablebooks/mdformat) plugin for [GitHub "Alerts"](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts). "Alerts" are a block quote variation of admonitions that were proposed in this [community discussion](https://github.com/orgs/community/discussions/16925) and are currently a separate extension of the [GFM (GitHub-Flavored Markdown) syntax](https://github.github.com/gfm).

**Metadata**

* homepage URL: https://pypi.org/project/mdformat_gfm_alerts
* source URL: https://github.com/KyleKing/mdformat-gfm-alerts
* license: mit
* platforms: all (python)

This package will allow `mdformat` to properly process the admonition in md files as written on GitHub. Without the plugin `mdformat` will escape the square brackets in the admonitions.

```
> [!NOTE] 
> Some note
```
will become the following without the `mdformat-gfm-plugin`:
```
> \[!NOTE\] 
> Some note
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by Sigmanificient**
## Description of changes

Fixes https://github.com/NixOS/nixpkgs/issues/328644

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/mdformat-gfm-alerts/default.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  dependencies = [
+    mdformat
+    mdit-py-plugins
+  ];
+
+  nativeCheckInputs = [ pytestCheckHook ];
+
+  disabledTests = [
+  ];
+
```
Comment (fabaff) [inline pkgs/development/python-modules/mdformat-gfm-alerts/default.nix:33]: Can be removed.

```suggestion
```

File: pkgs/development/python-modules/mdformat-gfm-alerts/default.nix [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  pythonImportsCheck = [ "mdformat_gfm_alerts" ];
+
+  meta = with lib; {
+    description = "Format 'GitHub Markdown Alerts,' which use blockquotes to render admonitions";
+    homepage = "https://github.com/KyleKing/mdformat-gfm-alerts";
```
Comment (fabaff) [inline pkgs/development/python-modules/mdformat-gfm-alerts/default.nix:38]:
```suggestion
    homepage = "https://github.com/KyleKing/mdformat-gfm-alerts";
    changelog = "https://github.com/KyleKing/mdformat-gfm-alerts/releases/tag/v${version}";
```

**Comment by sudoforge** [comment]
I tested this in a dev shell via an additive input:

```
{{
  description = "mdformat-gfm-alerts test";

  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/9bd88f8d981b1abb1cbb385fbd731a106393acb0";
    flake-utils.url = "github:numtide/flake-utils";
  };

  outputs =
    {
      self,
      flake-utils,
      nixpkgs,
    }:
    flake-utils.lib.eachDefaultSystem (
      system:
      let
        pkgs = nixpkgs.legacyPackages.${system};
      in
      {
        devShell = pkgs.mkShell {
          packages = with pkgs; [
            (mdformat.withPlugins (p: with p; [
              mdformat-gfm
              mdformat-gfm-alerts
            ]))
          ];
        };
      }
    );
}
```

Yet, it doesn't seem to work; GFM alerts are still being escaped:

```
➜ mdformat --number README.md

➜ git diff | rg NOTE
-> [!NOTE]
+> \[!NOTE\]
```

Am I setting this up correctly? I'm admittedly not too familiar with how `withPlugins` works.

File: pkgs/development/python-modules/mdformat-gfm-alerts/default.nix [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    mdformat
+    mdit-py-plugins
+  ];
+
+  nativeCheckInputs = [ pytestCheckHook ];
+
+  pythonImportsCheck = [ "mdformat_gfm_alerts" ];
+
+  meta = with lib; {
+    description = "Format 'GitHub Markdown Alerts,' which use blockquotes to render admonitions";
```
Comment (Aleksanaa) [inline pkgs/development/python-modules/mdformat-gfm-alerts/default.nix:34]:
```suggestion
    description = "Format 'GitHub Markdown Alerts', which use blockquotes to render admonitions";
```

File: pkgs/development/python-modules/mdformat-gfm-alerts/default.nix [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  dependencies = [
+    mdformat
+    mdit-py-plugins
+  ];
+
+  nativeCheckInputs = [ pytestCheckHook ];
+
+  pythonImportsCheck = [ "mdformat_gfm_alerts" ];
+
+  meta = with lib; {
```
Comment (Aleksanaa) [inline pkgs/development/python-modules/mdformat-gfm-alerts/default.nix:33]:
```suggestion
  meta = {
```

File: pkgs/development/python-modules/mdformat-gfm-alerts/default.nix [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    repo = "mdformat-gfm-alerts";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-2EYdNCyS1LxcEnCXkOugAAGx5XLWV4cWTNkXjR8RVQo=";
+  };
+
+  build-system = [ flit-core ];
+  dependencies = [
+    mdformat
+    mdit-py-plugins
+  ];
```
Comment (Aleksanaa) [inline pkgs/development/python-modules/mdformat-gfm-alerts/default.nix:27]:
```suggestion
  build-system = [ flit-core ];

  dependencies = [
    mdformat
    mdit-py-plugins
  ];
```

**Comment by Sigmanificient** [comment]
@sudoforge II am not sure why, but I was able to get it loaded by commenting out `mdformat-gfm`

**Comment by sudoforge** [comment]
@Sigmanificient said:
> @sudoforge I am not sure why, but I was able to get it loaded by commenting out `mdformat-gfm`

okay... this is weird. i can confirm that removing `mdformat-gfm` makes `mdformat-gfm-alerts` show up in the `--help` output:

```
➜ mdformat --help | rg plugins
Installed plugins: mdformat_gfm_alerts: 1.0.1
```

... but now `mdformat-tables` isn't loaded.

**Comment by sudoforge** [comment]
i think perhaps mdformat-gfm is picked up first, and... for... reasons, mdformat-gfm-alerts isn't. and once mdformat-gfm is removed, i have no clue why mdformat-tables wouldn't be loaded.

could be an upstream issue. i only started using mdformat recently, but i haven't encountered this with other plugins.

**Comment by sudoforge** [comment]
probably not an upstream issue.

inverting the order of the plugins in the `withPlugins` array (such that `mdformat-tables` comes before `mdformat-gfm-alerts` causes `mdformat_tables` to be loaded, but not `mdformat_gfm_alerts`:

```
➜ mdformat --help | rg plugins
Installed plugins: mdformat_tables: 0.4.1
```

#### `flake.nix`

```
{
  description = "mdformat-gfm-alerts test";

  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/1bca03dfe35b3a9a2b9442ba7c04d521abc9d055";
    flake-utils.url = "github:numtide/flake-utils";
  };

  outputs =
    {
      self,
      flake-utils,
      nixpkgs,
    }:
    flake-utils.lib.eachDefaultSystem (
      system:
      let
        pkgs = nixpkgs.legacyPackages.${system};
      in
      {
        devShell = pkgs.mkShell {
          packages = with pkgs; [
            (mdformat.withPlugins (p: with p; [
              mdformat-tables
              mdformat-gfm-alerts
            ]))
          ];
        };
      }
    );
}
```

**Review by sudoforge — APPROVED** [review]
as mentioned in the comments, there's some odd behavior when adding multiple plugins, but this is likely due to issues within `withPlugins`, or how the application sources plugins.

i'm approving this, because i don't believe that these findings should block this PR.

**Pull request merged**